### ISSUE TITLE
Using closures as registered plugins

### ISSUE THREAD

**Issue by bertptrs** [smarty-php/smarty#273]
Currently, `Smarty::registerPlguin` allows you to supply a Closure (since it is a callable) but when using it in a template it does not compile, resulting in the error `Object of class Closure could not be converted to string`.

This is because (for functions, other types of plugins behave similarly), it hits the following code segment:

``` php
if (!is_array($function)) {
    $output = "{$function}({$_params},\$_smarty_tpl)";
} elseif (is_object($function[ 0 ])) {
    $output =
        "\$_smarty_tpl->smarty->registered_plugins[Smarty::PLUGIN_FUNCTION]['{$tag}'][0][0]->{$function[1]}({$_params},\$_smarty_tpl)";
} else {
    $output = "{$function[0]}::{$function[1]}({$_params},\$_smarty_tpl)";
}
```

The first case gets executed, since a Closure is not an array. This could be solved by adding the following case:

``` php
if ($function instanceof Closure) {
    $output = "\$_smarty_tpl->smarty->registered_plugins[Smarty::PLUGIN_FUNCTION]['{$tag}'][0]({$_params},\$_smarty_tpl)";
} elseif (!is_array($function)) {
// And so forth
```

Could this be supported? I could probably write a patch works for it. This is backwards compatible to at least PHP 5.2, as the class does not have to be loaded (or even exist) for the operator to work.

**Comment by bertptrs** [comment]
Additionally, the calling syntax for closures is valid for the other types as well, so the branching structure might be removed altogether.

**Comment by fox34** [comment]
You can use this patch as temporary workaround. Not perfect, but doing the job.
Derived from this patch for older smarty versions: https://gist.github.com/za-creature/1761969

[Smarty-v3.1.30.registerPlugin.closure.txt](https://github.com/smarty-php/smarty/files/424821/Smarty-v3.1.30.registerPlugin.closure.txt)

**Comment by uwetews** [comment]
The fix is now in the master branch and will later be included in 3.1.32
